elastio-nat-provision-lambda/lambda.py

```diff
@@ -0,0 +1,348 @@
+            'describe_stacks',
+            'Stacks',
+            StackName=stack_name,
+        )
+        return stacks[0]['StackStatus']
+    except cfn.exceptions.ClientError as e:
+        if 'does not exist' in str(e):
+            print(f"Stack with a name {stack_name} does not exist.")
+        else:
+            print(f"Error describing stack {stack_name}: {e}")
+            print(f"Assuming the stack {stack_name} does not exist.")
+        return None
+
+
+def is_stack_deployed(stack_name):
+    stack_status = get_stack_status(stack_name)
+
+    if stack_status == 'DELETE_IN_PROGRESS':
+        print(f"The stack {stack_name} is being deleted, waiting until the deletion completes.")
+        while True:
+            time.sleep(5)
+            stack_status = get_stack_status(stack_name)
+            if stack_status != 'DELETE_IN_PROGRESS':
+                break
+
+    if stack_status is None:
+        return False
+
+    print(f"Stack {stack_name} is deployed and has status {stack_status}.")
+    return True
+
+
+def deploy_nat_stack(stack_name, subnet_id, route_table_id):
+    response = cfn.create_stack(
+        StackName=stack_name,
+        TemplateURL=NAT_CFN_TEMPLATE_URL,
+        OnFailure='DELETE',
+        Parameters=[
+            {
+                'ParameterKey': 'SubnetId',
+                'ParameterValue': subnet_id
+            },
+            {
+                'ParameterKey': 'RouteTableId',
+                'ParameterValue': route_table_id
+            },
+        ],
+    )
+    print(f"Stack creation initiated for {stack_name}: {response}")
+
+
+def choose_subnet_for_nat(subnets, public_subnets_ids, instance_subnet_id):
+    instance_az = subnets[instance_subnet_id]['AvailabilityZone']
+
+    for subnet_id, subnet in subnets.items():
+        if subnet_id not in public_subnets_ids:
+            continue
+        if subnet['AvailabilityZone'] != instance_az:
+            continue
+        return subnet_id
+
+
+def get_public_subnets(subnet_to_route_table, route_tables):
+    for subnet_id, rt_id in subnet_to_route_table.items():
+        route_table = route_tables[rt_id]
+
+        route = get_all_traffic_route(route_table)
+        if route is None:
+            continue
+        if not route['State'] == 'active':
+            continue
+
+        gw_id = route.get('GatewayId')
+        if gw_id is None or not gw_id.startswith('igw-'):
+            continue
+
+        yield subnet_id
+
+
+def get_all_traffic_route(route_table):
+    for route in route_table['Routes']:
+        if route.get('DestinationCidrBlock') == '0.0.0.0/0':
+            return route
+    return None
+
+
+def cleanup_nat(current_instance_id, event_time):
+    elastio_instances = {instance['InstanceId']: instance for instance in request(
+        ec2,
+        'describe_instances',
+        'Reservations[].Instances[]',
+        Filters=[{'Name': 'tag:elastio:resource', 'Values': ['true']}],
+    )}
+
+    active_instances_count = sum(
```

Comment:
The logic here doesn't take into account that elastio vault can be deployed into multiple AZ where there may be multiple NATs one per each AZ. For example, if there are no instances in one AZ but there are some in other AZ I suppose we need to make sure the NAT in the first AZ is deleted.